### ISSUE TITLE
Inconsistent behaviour regarding `//` in `Url`

### ISSUE THREAD

**Issue by jdrphillips** [lemonlabsuk/scala-uri#304]
**Description**

scala-uri behaves inconsistently when presented with double slashes in `Url`.

It automatically strips multiple slashes from urls when building:

```
scala> Url.parse("https://example.com") / "/"
res9: io.lemonlabs.uri.Url = https://example.com

scala> Url.parse("https://example.com") / "/" / "/"
res10: io.lemonlabs.uri.Url = https://example.com
```

but retains a trailing slash if the original string contains one at the end already:

```
scala> Url.parse("https://example.com/") / "/" / "/"
res11: io.lemonlabs.uri.Url = https://example.com/

scala> Url.parse("https://example.com/") / "foo"
res12: io.lemonlabs.uri.Url = https://example.com//foo
```

Note the `//foo` above in `res12`.

The first two pass an equality test - but not with the last:

```
scala> res9 == res10
res12: Boolean = true

scala> res11 == res10
res13: Boolean = false
```

## Expected Behaviour

I would expect `Url.parse("https://example.com/")` to strip the trailing slash from the URL before parsing, and pass equality checks with `res9` and `res10`. I would expect `Url.parse("https://example.com/") / "foo"` to not contain a double `//`

**Comment by theon** [comment]
Hi @jdrphillips. Thanks for raising this! Sorry for the delay getting back to you.

Totally agree that scala-uri needs normalisation of URLs like `https://example.com/` and `https://example.com` as RFC3986 discusses in 6.2.3 (quoted below) how these two URLs are equivalent. It actually says that `https://example.com/` should be used as the normal form, which is kind of annoying for solving the problem you highlight.

I'm thinking we could maybe make this configurable so users have the option to choose which normal form they want to use. Something like:

```
implicit val conf = UriConfig(pathNormalForm = RemoveTrailingSlashes) // Other options are `SingleTrailingSlash` and `Off`
Uri.parse("https://example.com/").toString // this is `https://example.com`
```

Let me know what you think.

> 6.2.3.  Scheme-Based Normalization
> 
>    The syntax and semantics of URIs vary from scheme to scheme, as
>    described by the defining specification for each scheme.
>    Implementations may use scheme-specific rules, at further processing
>    cost, to reduce the probability of false negatives.  For example,
>    because the "http" scheme makes use of an authority component, has a
>    default port of "80", and defines an empty path to be equivalent to
>    "/", the following four URIs are equivalent:
> 
>       http://example.com
>       http://example.com/
>       http://example.com:/
>       http://example.com:80/
> 
>    In general, a URI that uses the generic syntax for authority with an
>    empty path should be normalized to a path of "/".  Likewise, an
>    explicit ":port", for which the port is empty or the default for the
>    scheme, is equivalent to one where the port and its ":" delimiter are
>    elided and thus should be removed by scheme-based normalization.  For
>    example, the second URI above is the normal form for the "http"
>    scheme.

**Comment by jdrphillips** [comment]
No worries, thanks for getting back to me!

My only worry with that is the default behaviour of the library is to retain `//` inside URL paths, and `parse` accepts a default implicit parameter which would give this behaviour. While not _wrong_ it is immoral, and plenty of other systems don't handle `//` in url paths (which is how I found the issue in the first place).

It would be very difficult to impossible to detect when you have refactored things and lost your custom config along the way.

I would suggest flipping the behaviour and making the user have to supply a config in order to retain the current `/` behaviour - which is immoral so should take effort - and everyone else gets upgraded to a slicker experience automatically.

**Comment by jdrphillips** [comment]
But if your primary concern is behaving to the spec then your suggestion would be best I think. Maybe an additional `parse` method without a default implicit config?

**Comment by theon** [comment]
I'm happy for the default behaviour to remove trailing slashes and users having to opt-in to either have trailing slashes added (strictly following the spec) or leaving as-is.

I'm also thinking now that rather than normalizing automatically on `parse` (or on `toString` or any other time) it maybe is better to let the user decide when they want to normalize by providing a `normalize` method.

```scala
Url.parse("https://example.com/").normalize().toString // this is `https://example.com`
Url.parse("https://example.com").normalize(path = TrailingSlash).toString // this is `https://example.com/`

val url = Url.parse("https://example.com") / "" / "" // this is `https://example.com//`
url.normalize() // this is `https://example.com`
```

This could also allow for port normalization:

```scala
Url.parse("http://example.com:80/").normalize().toString // this is `https://example.com`
```

**Comment by jdrphillips** [comment]
Would `normalise` strip double `//` from within an already constructed path?

```scala
val u = Url.parse(http://example.com/) / "bar"
u.normalise() = ???
```

It would be the difference between normalising every url parsing in the system just to be sure, and just putting `normalise` on our encoders to external services (json, db, whatever)

**Comment by theon** [comment]
I think removing empty path segments (double slash) from the middle of paths would be a reasonable default for the `normalize` method. 

I've reread RFC3986 and it is only talking about the entire path being empty vs being a slash. [wikipedia](https://en.wikipedia.org/wiki/URI_normalization) categorises that as a "normalization that preserve semantics" whereas removing double slashes in the middle of a path as a "normalization that changes semantics". So I am thinking they should be treated separately by the `normalize` method.

```scala
def normalize(
  emptyPath = NoSlashEmptyPath, // can also be SlashEmptyPath or LeaveEmptyPath
  emptyPathSegments = RemoveEmptyPathSegments, // can also be LeaveEmptyPathSegments
  defaultPort = RemoveDefaultPort, // can also be ExplicitDefaultPort or LeaveDefaultPort
  ...
)
```

```scala
val u = Url.parse("http://example.com/") / "bar"
u.normalize() // This is "http://example.com/bar"
```

**Comment by jdrphillips** [comment]
👍 Sounds good!

**Comment by theon** [comment]
This has now been implemented and released in version `3.4.1`. Thanks to https://github.com/lemonlabsuk/scala-uri/pull/325

```scala
val url = Url.parse("https://example.com//a//b/c/d/")
val normalized = url.normalize(removeEmptyPathParts = true)
// val normalized: Url = https://example.com/a/b/c/d
```

`removeEmptyPathParts = true` will remove double slashes within a path and single trailing slashes at the end of the path. Let me know if you need anything else 👍
